### PR TITLE
Added Write Support in the PDS3Image.

### PR DESCRIPTION
**Pull request by bvnayak**
refs#45

### CODE REVIEW

**Comment by godber** [comment]
All of the sudden all Python 3 tests are erroring with:

```
urllib.error.HTTPError: HTTP Error 403: Forbidden
```

I went back and reran the previously passed `v0.3.0` test for 3.5 and it now fails.  Maybe the IP from which the 3+ tests are running is now being blocked by PDS due to repeating requests or something.

**Comment by godber** [comment]
We may have to mirror our test data ourselves.

**Comment by bvnayak** [comment]
Tasks to be performed in this ticket:

1. Write tests after adding 'save' support.
2. I swapped the byte-order in case of 3-band Images. I will fetch the byte order information from the label now and then will swap if its required.

File: tests/test_pds3file.py [hunk 1]
```diff
@@ -70,6 +70,42 @@ def test_bz2_pds3_1band_labels():
...
+    assert image.dtype == new_image.dtype
+    assert image.start_byte == new_image.start_byte
+    assert image.shape == new_image.shape
+    assert image.byte_order == new_image.byte_order
+    assert image.size == new_image.size
+
+
+def test_image_save_overwrite():
+    image = PDS3Image.open(filename)
+    image.save('Temp_Image.IMG')
```
Comment (bvnayak) [inline tests/test_pds3file.py]: Creating Temp Image first.

File: tests/test_pds3file.py [hunk 2]
```diff
@@ -70,6 +70,42 @@ def test_bz2_pds3_1band_labels():
...
+    assert image_temp.lines == new_image.lines
+    assert image_temp.samples == new_image.samples
+    assert image_temp.format == new_image.format
+    assert image_temp.pixel_type == new_image.pixel_type
+    assert image_temp.dtype == new_image.dtype
+    assert image_temp.start_byte == new_image.start_byte
+    assert image_temp.shape == new_image.shape
+    assert image_temp.byte_order == new_image.byte_order
+    assert image_temp.size == new_image.size
+    os.remove('Temp_Image.IMG')
```
Comment (bvnayak) [inline tests/test_pds3file.py]: Deleting the Temp Image at the end of test.

File: planetaryimage/image.py [hunk 3]
```diff
@@ -93,6 +93,29 @@ def get_nested_dict(item, key_list):
         """Get value from nested dictionary using a key list. """
         return reduce(lambda x, y: x[y], key_list, item)
 
+    def save(self, file_to_write=None, overwrite=False):
+        if overwrite:
+            file_to_write = self.filename
+        serial_label = dumps(self.label)
+        label_sz = len(serial_label)
+        image_pointer = int(label_sz / self.label['RECORD_BYTES']) + 1
```
Comment (bvnayak) [inline planetaryimage/image.py]: Using ceiling function here. (int(x/y) + 1)
1. Using math.ceil breaks TRAVIS CI for Py26 and Py 27.
2. range() does't accept argument from np.ceil().

File: planetaryimage/image.py [hunk 4]
```diff
@@ -292,7 +315,7 @@ def _parse_data(self, stream):
             data_stream.close()
 
     def _parse_band_sequential_data(self, stream):
-        if self.compression == 'none':
+        if self.compression is None:
```
Comment (bvnayak) [inline planetaryimage/image.py]: I changed default compression to `None` from 'none'.

**Comment by godber** [comment]
@bvnayak you can ignore these comments until next week, since I know you're studying.

I have a number of concerns with this PR:

* `save()` is implemented in `Image` but currently only handles saving PDS3 images.  What happens when a user tries to save an Isis cubefile?  Probably not the right thing.  So we can either leave `save()` up here and raise an exception when an Isis cubefile is being saved or we can move `save()` down to the `PDS3Image` class.  This implementation looks pretty PDS3 specific so maybe it should be moved down into `pds3image.py`.
* The `overwrite` keyword doesn't do what I had intended for it to do.  `save()` should test to see if a file matching `file_to_write`  is present, if it is present and `overwrite=False` then an exception should be raise with the message "File already exists: filename".  ONLY if `overwrite=True` should the file be overwritten.  The whole point is that `save()` should never overwrite an existing file without explicit permission.  This can probably be accomplished in the missing `else:` branch of the `if overwrite:` test.

Lastly, I've realized that many of our tests for pds3image are implemented incorrectly.  They should be testing `image.label` and `image.data`.  They mostly test the convenience properties that should maybe be considered internal and changed to `band` -> `_band`.  But at the very least, label values need to be tested explicitly since that's the interface users should be using.  Please modify your new tests here to include tests of `image.label` and `image.data`.  I will file another ticket to change the older tests.

**Comment by bvnayak** [comment]
Added the write support on the updated master branch. Closing this.

**Pull request closed without merging**